### PR TITLE
feat: ed25519 signature example

### LINKED ISSUE #682: examples utilizing misc. kinds of signing
The basic transfer example, only using a different signature type / means of signing. Perhaps https://deno.land/x/ed25519

Research involves looking into which test chains support a ed25519 scheme and how to run it w/ zombienet. 

implementation could involve modifying https://github.com/paritytech/capi/blob/main/crypto/Sr25519.ts to become a subclass that inherits a base class and adding a ED25519 subclass.

### PR DESCRIPTION
**Pull request by ryanleecode**
Resolves #682

## Changes
- ~~Updates the test user public keys list to also have ed25119 keys~~
- ~~Abstract Sr25519 into a Keypair class and adds Ed25519~~
- Adds transfer example using ED2519 signatures
- ~~Adds script for recreating test user public keys~~

## Notes
- ~~import map change is a requirement from https://github.com/paulmillr/noble-ed25519/tree/1.7.3~~
- ~~Not using latest version of noble-ed25519 b/c it doesn't work~~
- ~~Using older version of `main` because `main` is currently broken on linux machines. See #821~~

### CODE REVIEW

**Review by tjjfvi — CHANGES_REQUESTED** [review]
> - Not using latest version of noble-ed25519 b/c it doesn't work

Can you elaborate?

> - import map change is a requirement from https://github.com/paulmillr/noble-ed25519/tree/1.7.3

Can you use `https://esm.sh/@noble/ed25519` instead?

> - Adds script for recreating test user public keys

There is already a script for this: <https://github.com/paritytech/capi/blob/main/_tasks/generate_artifacts.ts#L30-L37>

> - Updates the test user public keys list to also have ed25119 keys
> - Abstract Sr25519 into a Keypair class and adds Ed25519

We don't use ed25519 anywhere else, so I'd like to keep the relevant logic contained to the example.

Instead of putting this logic into the crypto directory, can you write the signing logic etc. into the example? (I.e. without using a `Ed25519` class). It will then also be a good example of how the signing works under the hood.

Then, rather than add ed25519 test users, you can have an existential deposit from `alexa` to the ed25519 account (perhaps `eddie`), and then do a transfer from `eddie` to `billy` using the ed25519 signing.

**Comment by ryanleecode** [comment]
> Can you elaborate?

Running noble ed25519 on latest version gives me an error about sha512sync

```
vscode ➜ /workspaces/capi (feat/ed25519-signature-example) $ deno task run examples/ed25519_signature.ts
Task run deno task capi -- deno run -A -r=http://localhost:4646/ "examples/ed25519_signature.ts"
Task capi deno run -A main.ts --dbg "--" "deno" "run" "-A" "-r=http://localhost:4646/" "examples/ed25519_signature.ts"
Reusing existing Capi server at "http://localhost:4646/"
error: Uncaught (in promise) Error: etc.sha512Sync not set
const err = (m = ''): never => { throw new Error(m); }; // error helper, messes-up stack trace
                                       ^
    at err (https://deno.land/x/ed25519@2.0.0/index.ts:12:40)
    at sha512s (https://deno.land/x/ed25519@2.0.0/index.ts:193:47)
    at getExtendedPublicKey (https://deno.land/x/ed25519@2.0.0/index.ts:208:55)
    at Module.getPublicKey (https://deno.land/x/ed25519@2.0.0/index.ts:211:44)
    at Function.fromSecret (file:///workspaces/capi/crypto/Keypair.ts:32:26)
    at Sr25519.toEd25519 (file:///workspaces/capi/crypto/Keypair.ts:70:38)
    at file:///workspaces/capi/examples/ed25519_signature.ts:6:28
```

> Can you use https://esm.sh/@noble/ed25519 instead?

Same error as above

> There is already a script for this: https://github.com/paritytech/capi/blob/main/_tasks/generate_artifacts.ts#L30-L37

didn't see that mb

> We don't use ed25519 anywhere else, so I'd like to keep the relevant logic contained to the example.

sounds good

**Comment by tjjfvi** [comment]
> > Can you use https://esm.sh/@noble/ed25519 instead?
> 
> Same error as above

In that case, what about https://esm.sh/@noble/ed25519@1.7.3

File: examples/ed25519_signature.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import { Balances, users } from "westend_dev/mod.js"
+import { MultiAddress } from "westend_dev/types/sp_runtime/multiaddress.js"
+import * as ed from "../deps/ed25519.ts"
```
Comment (tjjfvi) [inline examples/ed25519_signature.ts:3]: Since this is only a dependency of the example, let's inline the url here and delete `deps/ed25519.ts`.

**Comment by ryanleecode** [comment]
> https://esm.sh/@noble/ed25519@1.7.3

```
vscode ➜ /workspaces/capi (feat/ed25519-signature-example) $ deno task run examples/ed25519_signature.ts
Task run deno task capi -- deno run -A -r=http://localhost:4646/ "examples/ed25519_signature.ts"
Task capi deno run -A main.ts --dbg "--" "deno" "run" "-A" "-r=http://localhost:4646/" "examples/ed25519_signature.ts"
Reusing existing Capi server at "http://localhost:4646/"
error: Uncaught (in promise) Error: utils.sha512Sync must be set to use sync methods
    at Y (https://esm.sh/v113/@noble/ed25519@1.7.3/deno/ed25519.mjs:2:10289)
    at tt (https://esm.sh/v113/@noble/ed25519@1.7.3/deno/ed25519.mjs:2:10444)
    at Object.Zt [as sign] (https://esm.sh/v113/@noble/ed25519@1.7.3/deno/ed25519.mjs:2:10813)
    at Object.sign (file:///workspaces/capi/examples/ed25519_signature.ts:16:20)
    at Codec._encode (file:///workspaces/capi/frame_metadata/Extrinsic.ts:63:40)
    at Codec._encode (https://deno.land/x/scale@v0.11.0/codecs/lenPrefixed.ts:14:14)
    at Codec.encodeAsync (https://deno.land/x/scale@v0.11.0/common/codec.ts:122:10)
    at RunMap.fn (file:///workspaces/capi/fluent/CodecRune.ts:9:26)
    at RunMap._evaluate (file:///workspaces/capi/rune/ValueRune.ts:150:40)
    at async RunMap.evaluate (file:///workspaces/capi/rune/Rune.ts:222:14)
```

File: examples/ed25519_signature.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import { Rune } from "../mod.ts"
+import { signature } from "../patterns/signature/polkadot.ts"
+
+const [alexa, billy, eddie] = await users(3)
+
+const ed25519PrivateKey = eddie.secretKey.slice(0, 32)
+const ed25519PublicKey = await ed.getPublicKey(ed25519PrivateKey)
```
Comment (tjjfvi) [inline examples/ed25519_signature.ts:10]:
```suggestion
const { alexa, billy } = await users()

const eddieSecretKey = crypto.getRandomValues(new Uint8Array(32))
const eddiePublicKey = await ed.getPublicKey(eddieSecretKey)
```

File: examples/ed25519_signature.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+const ed25519Eddie = Rune.rec({
```
Comment (tjjfvi) [inline examples/ed25519_signature.ts:12]:
```suggestion
const eddie = Rune.rec({
```

**Review by tjjfvi — COMMENTED** [review]
Looks a lot better; just a few nits

**Review by tjjfvi — APPROVED** [review]
Beautiful
image

**Pull request merged**